### PR TITLE
Add ex_aws as an optional dependency.

### PR DESCRIPTION
**Pull request by msmithstubbs**
Fixes a compiler error when `ex_aws` is included as a dependency but not yet compiled.

## Background
When compiling the `cloud_watch` dependency in a project using `ex_aws` I sometimes get a compiler error:

```
 ==> cloud_watch
could not compile dependency :cloud_watch, "mix compile" failed. You can recompile this dependency with "mix deps.compile cloud_watch", update it with "mix deps.update cloud_watch" or clean it with "mix deps.clean cloud_watch"
** (CondClauseError) no cond clause evaluated to a truthy value
      /opt/render/project/src/deps/cloud_watch/mix.exs:26: CloudWatch.Mixfile.application/0
```
This would only happen when compiling with clean deps and in a slower environment.

I understand this to be caused by the `ex_aws` not yet being compiled and available.

## Alternative solution

We could consider just having `aws` as the default dependency and allowing `ex_aws` to be used through an explicit config option:

```elixir
config :logger, CloudWatch,
  adapter: ExAws
```

### CODE REVIEW

**Comment by kuiro5** [comment]
Hey @msmithstubb, thanks for the PR! Tomorrow is American Thanksgiving, I'll try to take a deeper look at this over the weekend.

**Comment by kuiro5** [comment]
@msmithstubbs Looks like builds for Elixir versions 1.6-1.9 are failing on CI due to `telemetry` and `mime` not compiling on those versions. If we're able to get that sorted out, happy to get this merged.

**Comment by msmithstubbs** [comment]
Thanks for the feedback. `ex_aws` depends on `mime`, which only supports Elixir version 1.10 or greater.
Based on that I don't think I'll be able to keep `ex_aws` as an optional dependency, right? Any ideas I might have missed?

**Comment by msmithstubbs** [comment]
`mime` now compiles but still working on `telemetry`.

**Comment by kuiro5** [comment]
@msmithstubbs awesome progress! I can try to take a look at telemetry as well.

**Comment by msmithstubbs** [comment]
I've tracked this down to an error compiling telemetry on OTP releases < 21

The only thing I can think of is do add some version check log in `mix.exs`:

```elixir
 defp deps do
    [
      {:aws, "~> 0.5", optional: true},
      {:httpoison, ">= 0.11.1"},
      {:credo, "~> 1.4.0", only: :dev},
      {:mock, "~> 0.3.5", only: :test},
      {:ex_doc, "~> 0.22", only: :dev, runtime: false}
    ] ++ optional_ex_aws_deps()
  end

  # returns :ex_aws if OTP release >= 21.0
  def optional_ex_aws_deps do 
   ...
  end

```

What do you think @kuiro5?

File: mix.exs [hunk 1]
```diff
@@ -45,6 +45,9 @@ defmodule CloudWatch.Mixfile do
   defp deps do
     [
       {:aws, "~> 0.5", optional: true},
+      # Include mime for ex_aws; mime 2.x requires Elixir ~> 1.10
+      {:mime, "~> 1.2 or ~> 2.0", optional: true},
+      {:ex_aws, "~> 2.2", optional: true},
```
Comment (cichacz) [inline mix.exs:50]: Hello, can't we do the same with telemetry like with mime?
```suggestion
      {:ex_aws, "~> 2.2", optional: true},
      {:telemetry, "~> 0.4 or ~> 1.0"},
```

**Comment by kuiro5** [comment]
@msmithstubbs I'm going to close this in favor of #29. New version should be out tonight.

**Pull request closed without merging**